### PR TITLE
Surveys crud

### PR DESCRIPTION
**Pull request by avalmas-programize**
Includes all survey and survey media endpoints: (baseURL: /surveys)
- POST /: Create a new survey. The new survey entity is returned
- POST /:id/media/: Create a new survey media on survey with id ":id". The new survey media entity is returned
- POST /upload/: Upload a new image or video for the survey. The public URL of the media is returned
- GET /reefs/:id/: Get all surveys related to the reef
- GET /:id/: Get survey with its survey media grouped by survey points (reefPointOfInterest)
- PUT /:id/ Update survey. Update-able fields: diveData, weatherConditions, comments
- PUT /media/:id/ Update survey media. Update-able fields: featured, hidden, observations, comments, poiId (Survey Point)
- DELETE /:id/ Delete survey. Related survey media are deleted also (CASCADE). Media uploaded in the cloud are not deleted.
- DELETE /media/:id/ Delete survey media. Media uploaded in the cloud are not deleted.

### CODE REVIEW

File: packages/api/migration/1597927323620-UpdateSurveysTable.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
+import { MigrationInterface, QueryRunner } from 'typeorm';
+
+export class UpdateSurveysTable1597927323620 implements MigrationInterface {
+  name = 'UpdateSurveysTable1597927323620';
+
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(`ALTER TABLE "survey" DROP COLUMN "observations"`);
+    await queryRunner.query(`ALTER TABLE "survey" DROP COLUMN "upload_date"`);
+    await queryRunner.query(
+      `CREATE TYPE "survey_weather_conditions_enum" AS ENUM('calm', 'wavy', 'stormy')`,
```
Comment (ericboucher) [inline packages/api/migration/1597927323620-UpdateSurveysTable.ts:10]: let's call these
"calm", "waves", "storm"

File: packages/api/src/reefs/reefs.entity.ts [hunk 2]
```diff
@@ -69,4 +72,7 @@ export class Reef {
 
   @OneToOne(() => DailyData, (latestDailyData) => latestDailyData.reef)
   latestDailyData?: DailyData;
+
+  @OneToMany(() => Survey, (survey) => survey.reef)
+  surveys?: Survey;
```
Comment (ericboucher) [inline packages/api/src/reefs/reefs.entity.ts:77]: shouldn't this be a list Survey[] ?

File: packages/api/src/surveys/surveys.entity.ts [hunk 3]
```diff
@@ -7,36 +7,56 @@ import {
...
+  Calm = 'calm',
+  Wavy = 'wavy',
+  Stormy = 'stormy',
+}
 
 @Entity()
 export class Survey {
   @PrimaryGeneratedColumn()
-  id: number;
+  id?: number;
```
Comment (ericboucher) [inline packages/api/src/surveys/surveys.entity.ts:27]: Why this change? Other entities don't have the id optional

File: packages/api/src/surveys/surveys.service.ts [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+        'surveyMedia.id',
+        'surveyMedia.featured',
+        'surveyMedia.type',
+        'surveyMedia.observations',
+        'surveyMedia.comments',
+      ])
+      .addSelect(['poi.id', 'poi.imageUrl', 'poi.name'])
+      .getMany();
+
+    if (!survey) {
```
Comment (ericboucher) [inline packages/api/src/surveys/surveys.service.ts:106]: I think this could come right after:
```
const survey = await this.surveyRepository.findOne(surveyId);
```

**Comment by ericboucher** [comment]
I updated the code with my recommendations (please check @avalmas-programize) and deployed the backend on staging at https://staging-dot-ocean-systems.uc.r.appspot.com

File: packages/api/src/surveys/surveys.entity.ts (hunk 3 above)
Comment (avalmas-programize) [reply packages/api/src/surveys/surveys.entity.ts]: Since id is a `primary generated column`, it is optional as it is generated automatically by the DB.
However since all other entities have not make the id optional I will delete this change to have a uniform approach.

File: packages/api/src/reefs/reefs.entity.ts (hunk 2 above)
Comment (avalmas-programize) [reply packages/api/src/reefs/reefs.entity.ts]: Yeah correct. Fixed

File: packages/api/migration/1597927323620-UpdateSurveysTable.ts (hunk 1 above)
Comment (avalmas-programize) [reply packages/api/migration/1597927323620-UpdateSurveysTable.ts]: Enum changed
